h/security/acl.py

```diff
@@ -78,3 +78,71 @@ def _for_group(cls, group):
             yield Allow, group.creator.userid, Permission.Group.ADMIN
             yield Allow, group.creator.userid, Permission.Group.MODERATE
             yield Allow, group.creator.userid, Permission.Group.UPSERT
+
+    @classmethod
+    def for_annotation(cls, annotation, group, allow_read_on_delete=False):
+        """Return a Pyramid ACL for this annotation.
+
+        :param annotation: Annotation in question
+        :param group: Group associated with the annotation (if any)
+        :param allow_read_on_delete: Grant READ permissions on deleted
+            annotations.
+        """
+        yield from cls._for_annotation(annotation, group, allow_read_on_delete)
+
+        yield DENY_ALL
+
+    @classmethod
+    def _for_annotation(cls, annotation, group, allow_read_on_delete):
+        # If the annotation has been deleted, nobody has any privileges on it
+        # any more.
+        if annotation.deleted and not allow_read_on_delete:
+            return
+
+        if annotation.shared:
+            # You can read an annotation if you can read the group it's in
+            yield from cls._map_acls(
+                ACL.for_group(group),
+                {Permission.Group.READ: Permission.Annotation.READ},
+            )
+        else:
+            yield Allow, annotation.userid, Permission.Annotation.READ
+
+        if annotation.deleted:
```

Comment:
Is this not handled already on `if annotation.deleted and not allow_read_on_delete:`?

Reply:
No because that requires "allow_read_on_delete" to be False. If it's true we get this far and then quit.

Reply:
And I guess it doesn't make sense to flag already deleted annotations so that's why it's in the middle there :+1: